File: packages/earthsci-toolkit/src/function-tables-lowering.test.ts
/**
 * Bit-equivalent table_lookup → interp.* lowering harness (esm-lhm).
 *
 * For each conformance fixture under `tests/conformance/function_tables/`,
 * load the file, walk the model equations, lower every `table_lookup`
 * node to the structurally-equivalent inline-`const` `interp.linear` /
 * `interp.bilinear` invocation prescribed by esm-spec §9.5.3, and assert
 * IEEE-754 binary64 agreement with the equivalent hand-written
 * inline-`const` lookup at the §9.2 tolerance contract (`abs: 0, rel: 0`,
 * non-FMA reference path).
 *
 * Both arms drive the same `dispatchClosedFunction` evaluator; the harness
 * catches lowering-side mistakes (wrong output slice, swapped axis order,
 * dropped input expression) by computing the reference value from the raw
 * `function_tables` block independently of the parsed `table_lookup` node.
 */
import { describe, it, expect } from 'vitest'
import { readFileSync } from 'node:fs'
import { resolve } from 'node:path'
import { load } from './parse.js'
import { dispatchClosedFunction } from './registered_functions.js'

const FIXTURES_ROOT = resolve(__dirname, '..', '..', '..', 'tests', 'conformance', 'function_tables')

function bitEq(a: number, b: number): boolean {
  const buf = new ArrayBuffer(16)
  const f = new Float64Array(buf)
  f[0] = a
  f[1] = b
  const u = new BigUint64Array(buf)
  return u[0] === u[1]
}

function loadFixture(scenario: string): any {
  const path = resolve(FIXTURES_ROOT, scenario, 'fixture.esm')
  return load(readFileSync(path, 'utf-8'))
}

interface Var {
  type?: string
  default?: number
}

function resolveAxisValue(expr: unknown, vars: Record<string, Var>): number {
  if (typeof expr === 'number') return expr
  if (typeof expr === 'string') {
    const v = vars[expr]
    if (v?.default === undefined) {
      throw new Error(`variable ${expr} has no default`)
    }
    return v.default
  }
  throw new Error(`complex axis input not exercised: ${JSON.stringify(expr)}`)
}

function resolveOutputIndex(node: any, outputs: string[] | undefined): number {
  if (node.output === undefined || node.output === null) return 0
  if (typeof node.output === 'number') return node.output
  if (typeof node.output === 'string') {
    if (!outputs) throw new Error('string output requires table.outputs')
    const idx = outputs.indexOf(node.output)
    if (idx < 0) throw new Error(`output ${node.output} not found`)
    return idx
  }
  throw new Error('table_lookup.output must be int or string')
}

function slice1d(data: any, idx: number, hasOutputs: boolean): number[] {
  return (hasOutputs ? data[idx] : data).map((v: any) => Number(v))
}
function slice2d(data: any, idx: number, hasOutputs: boolean): number[][] {
  const rows = hasOutputs ? data[idx] : data
  return rows.map((r: any) => r.map((v: any) => Number(v)))
}

function lowerAndEvaluate(node: any, file: any, vars: Record<string, Var>): number {
  expect(node.op).toBe('table_lookup')
  expect(node.args).toEqual([])
  const table = file.function_tables[node.table]
  const kind = (table.interpolation ?? 'linear') as string
  const outIdx = resolveOutputIndex(node, table.outputs)
  const hasOutputs = Array.isArray(table.outputs)

  if (kind === 'linear' && table.axes.length === 1) {
    const axis = table.axes[0]
    const slice = slice1d(table.data, outIdx, hasOutputs)
    const x = resolveAxisValue(node.axes[axis.name], vars)
    return dispatchClosedFunction('interp.linear', [
      slice,
      axis.values.map(Number),
      x,
    ])
  }
  if (kind === 'bilinear' && table.axes.length === 2) {
    const ax = table.axes[0]
    const ay = table.axes[1]
    const slice = slice2d(table.data, outIdx, hasOutputs)
    const x = resolveAxisValue(node.axes[ax.name], vars)
    const y = resolveAxisValue(node.axes[ay.name], vars)
    return dispatchClosedFunction('interp.bilinear', [
      slice,
      ax.values.map(Number),
      ay.values.map(Number),
      x,
      y,
    ])
  }
  throw new Error(`unsupported lowering: kind=${kind} axes=${table.axes.length}`)
}

function referenceInlineConst(
  tableId: string,
  output: string | null,
  outputIdxInt: number | null,
  axisInputs: Array<[string, string]>,
  file: any,
  vars: Record<string, Var>,
): number {
  const table = file.function_tables[tableId]
  const hasOutputs = Array.isArray(table.outputs)
  let idx: number
  if (output !== null) {
    idx = (table.outputs as string[]).indexOf(output)
  } else if (outputIdxInt !== null) {
    idx = outputIdxInt
  } else {
    idx = 0
  }
  const kind = (table.interpolation ?? 'linear') as string

  if (kind === 'linear') {
    const axis = table.axes[0]
    const [axName, varName] = axisInputs[0]
    expect(axName).toBe(axis.name)
    const slice = slice1d(table.data, idx, hasOutputs)
    const x = vars[varName].default!
    return dispatchClosedFunction('interp.linear', [slice, axis.values.map(Number), x])
  }
  if (kind === 'bilinear') {
    const [ax, ay] = table.axes
    const [[axn0, vn0], [axn1, vn1]] = axisInputs
    expect(axn0).toBe(ax.name)
    expect(axn1).toBe(ay.name)
    const slice = slice2d(table.data, idx, hasOutputs)
    return dispatchClosedFunction('interp.bilinear', [
      slice,
      ax.values.map(Number),
      ay.values.map(Number),
      vars[vn0].default!,
      vars[vn1].default!,
    ])
  }
  throw new Error(`unsupported reference kind: ${kind}`)
}

describe('table_lookup → interp.* lowering bit-equivalence (esm-spec §9.5.3)', () => {
  it('linear fixture lowering matches the inline-const reference bit-for-bit', () => {
    const file = loadFixture('linear')
    const model = file.models!.M
    const vars = model.variables as Record<string, Var>
    const node = model.equations[0].rhs

    const lowered = lowerAndEvaluate(node, file, vars)
    const reference = referenceInlineConst(
      'sigma_O3_298',
      null,
      null,
      [['lambda_idx', 'lambda']],
      file,
      vars,
    )
    expect(bitEq(lowered, reference)).toBe(true)

    // Sanity: lambda=4.5 → i=3, w=0.5 → t3 + 0.5*(t4-t3)
    const expected = 8.7e-18 + 0.5 * (7.9e-18 - 8.7e-18)
    expect(bitEq(lowered, expected)).toBe(true)
  })

  it('bilinear fixture lowering matches the inline-const reference for both outputs', () => {
    const file = loadFixture('bilinear')
    const model = file.models!.M
    const vars = model.variables as Record<string, Var>

    const node0 = model.equations[0].rhs
    const lowered0 = lowerAndEvaluate(node0, file, vars)
    const reference0 = referenceInlineConst(
      'F_actinic',
      'NO2',
      null,
      [
        ['P', 'P_atm'],
        ['cos_sza', 'cos_sza'],
      ],
      file,
      vars,
    )
    expect(bitEq(lowered0, reference0)).toBe(true)

    const node1 = model.equations[1].rhs
    const lowered1 = lowerAndEvaluate(node1, file, vars)
    const reference1 = referenceInlineConst(
      'F_actinic',
      null,
      1,
      [
        ['P', 'P_atm'],
        ['cos_sza', 'cos_sza'],
      ],
      file,
      vars,
    )
    expect(bitEq(lowered1, reference1)).toBe(true)

    // Sanity: P=100, cos_sza=0.5 sits on the (1,1) interior knot.
    expect(lowered0).toBe(1.6) // NO2: data[0][1][1]
    expect(lowered1).toBe(2.6) // O3:  data[1][1][1]
  })

  it('roundtrip fixture lowering matches its inline-const companion bit-for-bit', () => {
    const file = loadFixture('roundtrip')
    const model = file.models!.M
    const vars = model.variables as Record<string, Var>
    const node = model.equations[0].rhs
    const lowered = lowerAndEvaluate(node, file, vars)

    // Eq 1 carries the equivalent inline-const interp.linear call by hand.
    const inline: any = model.equations[1].rhs
    expect(inline.op).toBe('fn')
    expect(inline.name).toBe('interp.linear')
    const tableArg: any = inline.args[0]
    const axisArg: any = inline.args[1]
    expect(tableArg.op).toBe('const')
    expect(axisArg.op).toBe('const')
    const inlineVal = dispatchClosedFunction('interp.linear', [
      (tableArg.value as number[]).map(Number),
      (axisArg.value as number[]).map(Number),
      resolveAxisValue(inline.args[2], vars),
    ])
    expect(bitEq(lowered, inlineVal)).toBe(true)
  })
})

File: packages/EarthSciSerialization.jl/test/function_tables_lowering_test.jl
# Bit-equivalent table_lookup → interp.* lowering harness (esm-lhm).
#
# For each conformance fixture under tests/conformance/function_tables/,
# load the file, walk the model equations, lower every table_lookup node
# to the structurally-equivalent inline-`const` interp.linear / interp.bilinear
# invocation prescribed by esm-spec §9.5.3, and assert IEEE-754 binary64
# agreement with the equivalent hand-written inline-`const` lookup at the
# §9.2 tolerance contract (abs: 0, rel: 0, non-FMA reference path).
#
# Both arms drive the same `evaluate_closed_function`; the harness catches
# lowering-side mistakes (wrong output slice, swapped axis order, dropped
# input expression) by computing the reference value from the raw
# `function_tables` block independently of the parsed table_lookup node.

using Test
using EarthSciSerialization
using EarthSciSerialization: evaluate_closed_function, OpExpr, NumExpr,
    IntExpr, VarExpr

const FUNCTION_TABLES_FIXTURES_ROOT = abspath(joinpath(@__DIR__, "..", "..", "..",
    "tests", "conformance", "function_tables"))

bit_eq(a::Float64, b::Float64)::Bool = reinterpret(UInt64, a) == reinterpret(UInt64, b)

resolve_axis_value(e::NumExpr, vars) = Float64(e.value)
resolve_axis_value(e::IntExpr, vars) = Float64(e.value)
function resolve_axis_value(e::VarExpr, vars)
    var = vars[e.name]
    @assert var.default !== nothing "variable $(e.name) has no default"
    Float64(var.default)
end
resolve_axis_value(e::OpExpr, _) =
    error("complex axis input expression not exercised: $(e.op)")

function resolve_output_index(node::OpExpr, outputs)
    out = node.output
    out === nothing && return 1  # 1-based for Julia
    if out isa Integer
        return Int(out) + 1  # convert 0-based wire index to 1-based
    elseif out isa AbstractString
        @assert outputs !== nothing "string output requires table.outputs"
        idx = findfirst(==(String(out)), outputs)
        @assert idx !== nothing "output name $out not in $outputs"
        return idx
    end
    error("table_lookup.output must be int or string, got $(typeof(out))")
end

slice_1d(data, idx, has_outputs) =
    Float64.(has_outputs ? data[idx] : data)

slice_2d(data, idx, has_outputs) =
    [Float64.(row) for row in (has_outputs ? data[idx] : data)]

function lower_and_evaluate(node::OpExpr, file, vars)
    @test node.op == "table_lookup"
    @test isempty(node.args)
    table = file.function_tables[node.table]
    kind = something(table.interpolation, "linear")
    out_idx = resolve_output_index(node, table.outputs)
    has_outputs = table.outputs !== nothing
    axes_map = node.table_axes::Dict{String,<:Any}

    if kind == "linear" && length(table.axes) == 1
        axis = table.axes[1]
        slice = slice_1d(table.data, out_idx, has_outputs)
        x = resolve_axis_value(axes_map[axis.name], vars)
        return Float64(evaluate_closed_function("interp.linear",
            Any[slice, Float64.(axis.values), x]))
    elseif kind == "bilinear" && length(table.axes) == 2
        ax = table.axes[1]; ay = table.axes[2]
        slice = slice_2d(table.data, out_idx, has_outputs)
        x = resolve_axis_value(axes_map[ax.name], vars)
        y = resolve_axis_value(axes_map[ay.name], vars)
        return Float64(evaluate_closed_function("interp.bilinear",
            Any[slice, Float64.(ax.values), Float64.(ay.values), x, y]))
    end
    error("unsupported lowering: kind=$kind axes=$(length(table.axes))")
end

function reference_inline_const(table_id::String, output, output_idx_int,
                                axis_inputs::Vector, file, vars)
    table = file.function_tables[table_id]
    has_outputs = table.outputs !== nothing
    if output !== nothing
        idx = findfirst(==(output), table.outputs)
        @assert idx !== nothing
    elseif output_idx_int !== nothing
        idx = output_idx_int + 1
    else
        idx = 1
    end
    kind = something(table.interpolation, "linear")
    if kind == "linear"
        axis = table.axes[1]
        ax_name, var_name = axis_inputs[1]
        @test ax_name == axis.name
        slice = slice_1d(table.data, idx, has_outputs)
        x = Float64(vars[var_name].default)
        return Float64(evaluate_closed_function("interp.linear",
            Any[slice, Float64.(axis.values), x]))
    elseif kind == "bilinear"
        ax = table.axes[1]; ay = table.axes[2]
        (axn0, vn0), (axn1, vn1) = axis_inputs[1], axis_inputs[2]
        @test axn0 == ax.name
        @test axn1 == ay.name
        slice = slice_2d(table.data, idx, has_outputs)
        return Float64(evaluate_closed_function("interp.bilinear",
            Any[slice, Float64.(ax.values), Float64.(ay.values),
                Float64(vars[vn0].default), Float64(vars[vn1].default)]))
    end
    error("unsupported reference kind: $kind")
end

@testset "table_lookup → interp.* lowering bit-equivalence (esm-spec §9.5.3)" begin
    @testset "linear fixture" begin
        path = joinpath(FUNCTION_TABLES_FIXTURES_ROOT, "linear", "fixture.esm")
        file = EarthSciSerialization.load(path)
        model = file.models["M"]
        vars = model.variables

        node = model.equations[1].rhs::OpExpr
        lowered = lower_and_evaluate(node, file, vars)
        reference = reference_inline_const("sigma_O3_298", nothing, nothing,
            [("lambda_idx", "lambda")], file, vars)
        @test bit_eq(lowered, reference)

        # Sanity: lambda=4.5 → interval [4,5], w=0.5
        expected = 8.7e-18 + 0.5 * (7.9e-18 - 8.7e-18)
        @test bit_eq(lowered, expected)
    end

    @testset "bilinear fixture" begin
        path = joinpath(FUNCTION_TABLES_FIXTURES_ROOT, "bilinear", "fixture.esm")
        file = EarthSciSerialization.load(path)
        model = file.models["M"]
        vars = model.variables

        node0 = model.equations[1].rhs::OpExpr
        lowered0 = lower_and_evaluate(node0, file, vars)
        reference0 = reference_inline_const("F_actinic", "NO2", nothing,
            [("P", "P_atm"), ("cos_sza", "cos_sza")], file, vars)
        @test bit_eq(lowered0, reference0)

        node1 = model.equations[2].rhs::OpExpr
        lowered1 = lower_and_evaluate(node1, file, vars)
        reference1 = reference_inline_const("F_actinic", nothing, 1,
            [("P", "P_atm"), ("cos_sza", "cos_sza")], file, vars)
        @test bit_eq(lowered1, reference1)

        @test lowered0 == 1.6
        @test lowered1 == 2.6
    end

    @testset "roundtrip fixture (table_lookup vs. inline-const companion)" begin
        path = joinpath(FUNCTION_TABLES_FIXTURES_ROOT, "roundtrip", "fixture.esm")
        file = EarthSciSerialization.load(path)
        model = file.models["M"]
        vars = model.variables

        node = model.equations[1].rhs::OpExpr
        lowered = lower_and_evaluate(node, file, vars)

        inline = model.equations[2].rhs::OpExpr
        @test inline.op == "fn"
        @test inline.name == "interp.linear"
        table_arg = inline.args[1]::OpExpr
        axis_arg = inline.args[2]::OpExpr
        @test table_arg.op == "const"
        @test axis_arg.op == "const"
        x = resolve_axis_value(inline.args[3], vars)
        inline_val = Float64(evaluate_closed_function("interp.linear",
            Any[Float64.(table_arg.value), Float64.(axis_arg.value), x]))
        @test bit_eq(lowered, inline_val)
    end
end
